### PR TITLE
Add array-like properties to `DeviceBuffer`

### PR DESCRIPTION
**Pull request by jakirkham**
Adds some common properties that array-like objects expose. This is useful when interacting with other array consuming libraries that use duck-typing in their functions.

### CODE REVIEW

File: python/rmm/_lib/device_buffer.pyx [hunk 1]
```diff
@@ -21,25 +23,41 @@ cdef class DeviceBuffer:
     def __len__(self):
         return self.size
 
+    @property
+    def dtype(self):
```
Comment (harrism) [inline python/rmm/_lib/device_buffer.pyx]: `rmm::device_buffer` is anuntyped memory buffer. So it doesn't make sense to treat it like a typed array.

Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx]: This is basically treating it as `unsigned char` ATM, which behaves similarly to checking type on a `memoryview` of `bytes` in Python (please see below).

```python
In [1]: m = memoryview(b"abc")                                                  

In [2]: m.format  # same as `|u1`
Out[2]: 'B'
```

Comment (kkraus14) [reply python/rmm/_lib/device_buffer.pyx]: What we're doing in `__cuda_array_interface__` is very much a workaround in my mind at the moment. Python Buffer protocol may have array-like properties, but that seems out of scope for just defining memory buffers, which is the goal of `DeviceBuffer`. If we want a thin array-like wrapper on top of `DeviceBuffer` that can wrap some things from `device_vector` as well then that would make more sense, but then we get into the territory of creating yet another GPU array library.

**Comment by kkraus14** [comment]
Closing this as out of scope.

File: python/rmm/_lib/device_buffer.pyx (hunk 1 above)
Comment (jakirkham) [reply python/rmm/_lib/device_buffer.pyx]: I wouldn't call using `__cuda_array_interface__` a workaround with `DeviceBuffer`. This is the same as supporting [the Python buffer protocol]( https://docs.python.org/3/c-api/buffer.html ) with `bytes` objects.

That said, I do see your point that we might not want `DeviceBuffer` to explicitly expose array properties.

**Pull request closed without merging**